vhost_user_block/src/lib.rs

```diff
@@ -171,7 +174,7 @@ impl VhostUserBlkThread {
                 Ok(request) => {
                     debug!("element is a valid request");
                     let status = match request.execute(
-                        &mut self.disk_image,
+                        &mut self.disk_image.lock().unwrap().deref_mut(),
```

Comment:
This is going to least to contention between each thread over the single shared `DiskFile`. Won't that remove almost all the benefits of the multithreading MQ? Also it looks like virtio-blk also has it shared too. @slp do you have any insight you could share with us on this?

Reply:
@rbradford I know, that's what I was talking about yesterday. I tried to identify where the need for mutable File is. It's located in `vm-virtio/src/block.rs` in the `execute()` function, and in particular it is due to the read/write from/to guest memory. Both `read_exact_from()` and `write_all_to()` expect a `&mut F` where F is either `Read` or `Write`.

Now that being said, I still think the PR makes sense to be merged as it right now since it implements multithread support for `vhost-user-blk`. As a follow up, if we can make execute() not require a `&mut DiskFile`, we will be able to improve the design further.

Reply:
So I see you have settled on the simpler approach for implementing multithreading by making `handle_event` immutable. Good call! (and I should really enable the notifications 😂)

As for the scalability of vhost-user-blk, @rbradford is right about the contention on `DiskFile` lock. To fix this, I think we can simply extend the `DiskFile` trait with a couple methods that would wrap `pread64` and `pwrite64` (we can't touch the file offset) and implement them in `RawFile`.

Perhaps this should be done in a different PR, since it would affect (and benefit!) `virtio-blk` too.

That said, I think I should also mention that this approach is always nice as a fallback, but ideally we should also implement support for asynchronous IO based on `libaio` and/or `io_uring`.

Reply:
>To fix this, I think we can simply extend the DiskFile trait with a couple methods that would wrap pread64 and pwrite64 (we can't touch the file offset) and implement them in RawFile.

But I thought the real bottleneck was vm-memory here since it requires `read_exact_from()` and `write_all_to()` to provide mutable `Read` or `Write` implementation. So I don't think the change to DiskFile is gonna be enough. Or did I misunderstand something?

>Perhaps this should be done in a different PR, since it would affect (and benefit!) virtio-blk too.

Yes I really think this can go into a dedicated PR.

>That said, I think I should also mention that this approach is always nice as a fallback, but ideally we should also implement support for asynchronous IO based on libaio and/or io_uring.

I'm not familiar with this, but doing some quick search on it, that sounds like a good long term plan to improve our virtio-blk performances.

Reply:
> > To fix this, I think we can simply extend the DiskFile trait with a couple methods that would wrap pread64 and pwrite64 (we can't touch the file offset) and implement them in RawFile.
> 
> But I thought the real bottleneck was vm-memory here since it requires `read_exact_from()` and `write_all_to()` to provide mutable `Read` or `Write` implementation. So I don't think the change to DiskFile is gonna be enough. Or did I misunderstand something?

Yeah, I forgot about that. In fact, I *think* when I last looked at this (6 months ago?) I decided to wait for Rust to implement a trait positional vectored IO which was supposed to arrive soon, but AFAIK it's still not there. Of course, vm-memory could implement a custom trait for this.

On the other hand, the libaio/io_uring won't require such change as completions can be serialized.

Reply:
Thanks for the feedback @slp 
Maybe it'd be worth adding your analysis to the new issue https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1037 tracking all the improvements we can come up with for parallelizing accesses to a block device.